tls/s2n_x509_validator.c

```diff
@@ -680,23 +691,53 @@ S2N_RESULT s2n_x509_validator_validate_cert_stapled_ocsp_response(struct s2n_x50
     OCSP_CERTID_free(cert_id);
     RESULT_GUARD_OSSL(ocsp_resp_find_status_res, S2N_ERR_CERT_UNTRUSTED);
 
-    uint64_t this_update = 0;
-    RESULT_GUARD(s2n_asn1_time_to_nano_since_epoch_ticks((const char *) thisupd->data,
-            (uint32_t) thisupd->length, &this_update));
+    uint64_t current_sys_time_nanoseconds = 0;
+    RESULT_GUARD(s2n_config_wall_clock(conn->config, &current_sys_time_nanoseconds));
+    if (sizeof(time_t) == 4) {
+        /* cast value to uint64_t to prevent overflow errors */
+        RESULT_ENSURE_LTE(current_sys_time_nanoseconds, (uint64_t) MAX_32_TIMESTAMP_NANOS);
+    }
+    /* convert the current_sys_time (which is in nanoseconds) to seconds */
+    time_t current_sys_time_seconds = (time_t) (current_sys_time_nanoseconds / ONE_SEC_IN_NANOS);
+
+    DEFER_CLEANUP(ASN1_GENERALIZEDTIME *current_sys_time = ASN1_GENERALIZEDTIME_set(0, current_sys_time_seconds), s2n_openssl_asn1_time_free_pointer);
+    RESULT_ENSURE_REF(current_sys_time);
 
-    uint64_t next_update = 0;
+    /**
+     * It is fine to use ASN1_TIME functions with ASN1_GENERALIZEDTIME structures
+     * From openssl documentation:
+     * It is recommended that functions starting with ASN1_TIME be used instead
+     * of those starting with ASN1_UTCTIME or ASN1_GENERALIZEDTIME. The
+     * functions starting with ASN1_UTCTIME and ASN1_GENERALIZEDTIME act only on
+     * that specific time format. The functions starting with ASN1_TIME will
+     * operate on either format.
+     * https://www.openssl.org/docs/man1.1.1/man3/ASN1_TIME_to_generalizedtime.html
+     *
+     * ASN1_TIME_compare has a much nicer API, but is not available in Openssl
+     * 1.0.1, so we use ASN1_TIME_diff.
+     */
+    int pday = 0;
+    int psec = 0;
+    RESULT_GUARD_OSSL(ASN1_TIME_diff(&pday, &psec, thisupd, current_sys_time), S2N_ERR_CERT_UNTRUSTED);
+    /* ensure that current_time is after or the same as "this update" */
+    RESULT_ENSURE(pday >= 0 && psec >= 0, S2N_ERR_CERT_INVALID);
+
+    /* ensure that current_time is before or the same as "next update" */
     if (nextupd) {
-        RESULT_GUARD(s2n_asn1_time_to_nano_since_epoch_ticks((const char *) nextupd->data,
-                (uint32_t) nextupd->length, &next_update));
+        RESULT_GUARD_OSSL(ASN1_TIME_diff(&pday, &psec, current_sys_time, nextupd), S2N_ERR_CERT_UNTRUSTED);
+        RESULT_ENSURE(pday >= 0 && psec >= 0, S2N_ERR_CERT_EXPIRED);
     } else {
-        next_update = this_update + DEFAULT_OCSP_NEXT_UPDATE_PERIOD;
+        /**
+         * if nextupd isn't present, assume that nextup is
+         * DEFAULT_OCSP_NEXT_UPDATE_PERIOD after thisupd. This means that if the
+         * current time was more than 0 days and DEFAULT_OCSP_NEXT_UPDATE_PERIOD
+         * seconds ahead of nextup, we consider it invalid. We already compared
+         * current_sys_time to thisupd, so reuse those values
```

Comment:
Some nits in this paragraph:
```suggestion
         * if nextupd isn't present, assume that nextupd is
         * DEFAULT_OCSP_NEXT_UPDATE_PERIOD after thisupd. This means that if the
         * current time is more than DEFAULT_OCSP_NEXT_UPDATE_PERIOD
         * seconds ahead of thisupd, we consider it invalid. We already compared
         * current_sys_time to thisupd, so reuse those values
```
Also, I'm not sure, but it might be more clear to use the actual RFC field names in comments/variable names? Like, this_update and next_update rather than thisupd/nextupd.

Reply:
The actual rfc field names are thisUpdate and nextUpdate, which I think are ugly, so I'd stick with the variable names? Opens to suggestions.